src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -1529,6 +1533,24 @@ class Checker(globalTable: SymbolTable, verbose: Boolean = false) {
 
     }
 
+    private def checkConstructors(constructors: Seq[Constructor], contract: Contract, table: ContractTable): Unit = {
+        if (constructors.isEmpty && table.stateLookup.nonEmpty) {
+            logError(contract, NoConstructorError(contract.name))
+        }
+
+        if (constructors.length > 1 && contract.isMain) {
+            logError(contract, MultipleConstructorsError(contract.name))
+        }
+
+        val constructorsByArgTypes = constructors.groupBy(c => c.args.map(_.typ))
```

Comment:
What if the types are not identical, but one is a subtype of the other? e.g. Foo@Owned, Foo@Unowned?